### DISCUSSION TITLE
help getting Q&A pairs back via API

### DISCUSSION THREAD

**Discussion by SilverSeed93** [oobabooga/text-generation-webui#6404] [Q&A]
Hi Everyone. 

I'm attempting to write a python script to talk to Text Generation WebUI, provide it a file for context, and get Q&A pairs back from the LLM. 

For the most part I don't know what I'm doing! I'm new at this and python both. I have managed to get this to work some of the time which is amazing in and of itself but the LLM will respond back with totally different raw responses and I'm having troubles parsing it into JSON.

Here is an example of when it works on file 1:
Assistant's raw response:
`
[{"question": "Who is the best tracker in Stargate SG-1?", "answer": "Master Bratac is the best tracker in the series Stargate SG-1, but if you are asking about intrateam skills than Tealc is the best tracker on the SG-1 team."},
`

Here is an example of when it breaks on the very next file:
Assistant's raw response:
`
Here are the questions based on the text:
What are the two main types of hyperdrive systems?
Assistant's raw response:
Here are the questions based on the text:
What are the two main types of hyperdrive systems?
`
Note: The second response provided 10 questions in non-json format and instead of answers just repeated the same 10 questions again not in json format.

Note: This is actually not about Stargate, just an example!

I've attempted many different prompts and provided an example of json I want attempting to force a structured response. I've tried find() to remove text that isn't json. I can't get the LLM to respond consistently. But... If the LLM doesn't respond properly I can't really parse properly!

Here is the bit I am not sure about:
`

    headers = {
        "Content-Type": "application/json"
    }
    
    prompt = (
        "Please create distinct question and answer pairs from the following content. "
        "Ensure the response is a valid JSON array formatted like this: "
        "[{\"question\": \"[Question]\", \"answer\": \"[Answer]\"}].\n"
        f"{markdown_content}\n"
    )

    history = [{"role": "user", "content": prompt}]
    data = {
        "mode": "chat",
        "messages": history,
        "max_tokens": 512,
    }
`

I am using:
Model: Meta-Llama-3-8B-Instruct-Q8_0.gguf
Loader: llama.cpp
The rest of the settings are all default including assistant's.

Could anyone help me with this?

If I can manage to get this working I'd like to train an LLM to answer questions based on this data as, factually, as possible. However, I just read this old post:
https://github.com/oobabooga/text-generation-webui/discussions/3038

Has anything changed since then? Can someone who has done this point me towards a few simple steps to look up so I can get started?

Thanks!

**Comment by Katehuuh** [comment]
With some luck you could try https://github.com/e-p-armstrong/augmentoolkit, if not try [prompt](https://huggingface.co/datasets/Nekochu/Tree-of-Web-Knowledge/blob/main/create_data.py#L71-L94) by [Trelis Research](https://youtu.be/JJ5mcdEIbj8?si=LwMieD3iR2915Mhm). If necessary you can force JSON output by using grammar file: `Parameters>Generation>Filter by loader>Load grammar from file (.gbnf)>json.gbnf` in WebUI, Or JSON via API:
 <details> <summary>Python chat example JSON constraint</summary>

```python
import requests
import json
url = "http://127.0.0.1:5000/v1/chat/completions"
headers = {"Content-Type": "application/json"}

# JSON grammar from "/user_data/grammars/json.gbnf"
json_grammar = '''
root ::= object
object ::= "{" ws (string ":" ws value ("," ws string ":" ws value)*)? ws "}"
value  ::= object | array | string | number | "true" | "false" | "null"
array  ::= "[" ws (value ("," ws value)*)? ws "]"
string ::= "\\"" [^"\\\\]* "\\""
number ::= "-"? [0-9]+ ("." [0-9]+)? ([eE] [+-]? [0-9]+)?
ws ::= [ \\t\\n\\r]*
'''

data = {
    "messages": [
        {
            "role": "user",
            "content": "Please create distinct question and answer pairs from the following content. Ensure the response is a valid JSON array formatted like this: [{\"question\": \"[Question]\", \"answer\": \"[Answer]\"}]. Machine learning is a subset of artificial intelligence that enables computers to learn and improve from experience without being explicitly programmed. It uses algorithms to identify patterns in data and make predictions or decisions. Common types include supervised learning, unsupervised learning, and reinforcement learning."
        }
    ],
    "grammar_string": json_grammar,
    "temperature": 0.1
}

response = requests.post(url, headers=headers, json=data)
result = response.json()
print("Proper JSON format:")
print(result['choices'][0]['message']['content'])

# Verify it's valid JSON
try:
    parsed = json.loads(result['choices'][0]['message']['content'])
    print("✅ Valid JSON:", parsed)
except json.JSONDecodeError:
    print("❌ Still invalid JSON")
```
Output:
```json
{"question": "What is machine learning?", "answer": "Machine learning is a subset of artificial intelligence that enables computers to learn and improve from experience without being explicitly programmed."}
```

Usable only with loader [llama.cpp](https://github.com/abetlen/llama-cpp-python?tab=readme-ov-file#json-schema-mode) [GBNF](https://github.com/ggml-org/llama.cpp/blob/master/grammars/README.md) implemented in [/modules/grammar](https://github.com/oobabooga/text-generation-webui/tree/main/modules/grammar) and also supported by `Transformers`, `ExLlamav2_HF` and `ExLlamav3_HF` since [v0.0.5](https://github.com/turboderp-org/exllamav3/releases/tag/v0.0.5).

Note that [function-call](https://github.com/oobabooga/text-generation-webui/wiki/12-%E2%80%90-OpenAI-API#toolfunction-calling-example) just uses regex:
https://github.com/oobabooga/text-generation-webui/blob/6338dc0051e02c29dcd4c9ee9d7fc6da20423cda/extensions/openai/utils.py#L93-L94

</details>

> If I can manage to get this working I'd like to train an LLM to answer questions based on this data as, factually, as possible. However, I just read this old post: #3038

For training, I recommend https://github.com/hiyouga/LLaMA-Factory QLoRA to save VRAM.

**Reply by SilverSeed93** [reply]
> With some luck you could try https://github.com/e-p-armstrong/augmentoolkit, if not try [prompt](https://huggingface.co/datasets/Nekochu/Tree-of-Web-Knowledge/blob/main/create_data.py#L71-L94) by [Trelis Research](https://youtu.be/JJ5mcdEIbj8?si=LwMieD3iR2915Mhm). If necessary you can force JSON output by using grammar file: `Parameters>Generation>Filter by loader>Load grammar from file (.gbnf)>json.gbnf`.
> 
> > If I can manage to get this working I'd like to train an LLM to answer questions based on this data as, factually, as possible. However, I just read this old post: #3038
> 
> For training, I recommend https://github.com/hiyouga/LLaMA-Factory QLoRA to save VRAM.

Thank you! I will have a look into all of this.
